{"cells":[{"cell_type":"markdown","metadata":{"id":"6_fcGwPVjjEn"},"source":["## Example protein design notebook\n","### Design of a protein minibinder\n","\n","Recent developments using AI for protein design and engineering have allowed bespoke protein designs to be generated *in silico* to achieve certain specific design criteria.\n","\n","In this case we are going to design a protein that will bind to the [p53](https://www.uniprot.org/uniprotkb/P04637/entry) tumour suppressor protein at the protein:DNA interface.\n","\n","This workflow briefly demonstrates the following fundamental protein design steps:\n","- Predict the structure of a target protein of interest using [Boltz2](https://www.rxrx.ai/boltz-2), an open-source implementation of [AlphaFold3](https://github.com/google-deepmind/alphafold3). This allows us to identify the protein:DNA interface.\n","- Design an artificial protein binder using [RFdiffusion](https://www.nature.com/articles/s41586-023-06415-8) that will bind to p53 at the protein:DNA interface.\n","- Generate a sequence using [proteinMPNN](https://www.science.org/doi/10.1126/science.add2187) that will fold into the desired binder structure.\n","- Analyse the generated sequences and select a preferred sequence.\n","- Predict the structure of the target protein with the designed binder again to make sure it folds into the correct structure and binds at the desired interface.\n","\n","This example code could readily be expanded to include:\n","- Generation of additional designs\n","- Generation of additional test sequences\n","- Additional analysis and screening steps to preselect sequences for wet lab screening.\n","- Generation of expression sequences and constructs.\n","- Generation of order forms and automation picklists."]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":173522,"status":"ok","timestamp":1753103766487,"user":{"displayName":"Peter Canning","userId":"16600209944252973081"},"user_tz":-60},"id":"Wi5fJ2cQjK5B","outputId":"14eb3610-3215-4ddd-e66b-c27687efd509"},"outputs":[{"output_type":"stream","name":"stdout","text":["\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m266.3/266.3 kB\u001b[0m \u001b[31m11.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n","  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n","  Installing backend dependencies ... \u001b[?25l\u001b[?25hdone\n","  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m118.2/118.2 kB\u001b[0m \u001b[31m8.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m75.8/75.8 kB\u001b[0m \u001b[31m5.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m92.7/92.7 kB\u001b[0m \u001b[31m6.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m392.5/392.5 kB\u001b[0m \u001b[31m19.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m154.5/154.5 kB\u001b[0m \u001b[31m9.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m825.4/825.4 kB\u001b[0m \u001b[31m35.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m34.9/34.9 MB\u001b[0m \u001b[31m19.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m103.0/103.0 kB\u001b[0m \u001b[31m6.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m94.0/94.0 kB\u001b[0m \u001b[31m6.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m70.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.6/2.6 MB\u001b[0m \u001b[31m68.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m17.7/17.7 MB\u001b[0m \u001b[31m72.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m139.1/139.1 kB\u001b[0m \u001b[31m11.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.4/56.4 kB\u001b[0m \u001b[31m3.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m963.5/963.5 kB\u001b[0m \u001b[31m39.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m126.2/126.2 kB\u001b[0m \u001b[31m9.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25h  Building wheel for fairscale (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n","  Building wheel for modelcif (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Building wheel for ihm (setup.py) ... \u001b[?25l\u001b[?25hdone\n","done.\n"]}],"source":["#@title Install dependencies (~2min)\n","\n","runtime = \"GPU(L4 or T4)\"\n","\n","import os\n","import subprocess\n","\n","print('Installing dependencies... ', end='')\n","dependencies = \"torch torchvision torchaudio numpy hydra-core pytorch-lightning py3dmol \"\n","dependencies += \"rdkit dm-tree requests pandas types-requests einops einx fairscale \"\n","dependencies += \"mashumaro modelcif wandb click pyyaml biopython scipy numba gemmi \"\n","dependencies += \"scikit-learn chembl_structure_pipeline \"\n","dependencies += \"cuequivariance_ops_cu12 cuequivariance_ops_torch_cu12 cuequivariance_torch\"\n","\n","if runtime == \"GPU(L4 or T4)\":\n","    precision = '32-true'\n","else:\n","    precision = 'bf16-true'\n","\n","subprocess.run(\"pip install ipywidgets torch torchvision torchaudio\", shell=True)\n","subprocess.run(\"git clone https://github.com/jwohlwend/boltz.git\", shell=True)\n","subprocess.run(f\"sed -i 's/bf16-mixed/{precision}/g' /content/boltz/src/boltz/main.py\", shell=True)\n","!pip install {dependencies} --quiet\n","subprocess.run(\"cd boltz; pip install --no-deps -e .\", shell=True)\n","\n","print('done.')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"q-zuDv0cnUxp","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1753103775680,"user_tz":-60,"elapsed":9172,"user":{"displayName":"Peter Canning","userId":"16600209944252973081"}},"outputId":"1d51d87b-82fe-41be-ea66-585c9492fd06"},"outputs":[{"output_type":"stream","name":"stdout","text":["This notebook will be very slow without a GPU. Request a GPU from Google Colab by going to Runtime > Change runtime type and selecting 'T4 GPU'\n"]}],"source":["import numpy\n","import torch\n","import pathlib\n","import boltz\n","import tempfile\n","import py3Dmol\n","\n","if not torch.cuda.is_available():\n","  raise RuntimeError(\"This notebook requires a GPU to run. Request a GPU from Google Colab by going to Runtime > Change runtime type and selecting 'T4 GPU'\")"]},{"cell_type":"markdown","metadata":{"id":"n3M7c7r231wt"},"source":["### Predict the target protein structure\n","\n","First we will predict the structure of p53 bound to DNA. In this case we can use the protein and DNA sequences used to solve the crystal structure of p53 bound to DNA, [1TSR](https://www.rcsb.org/structure/1TSR):"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":97645,"status":"ok","timestamp":1753103873323,"user":{"displayName":"Peter Canning","userId":"16600209944252973081"},"user_tz":-60},"id":"Izo-o2U-pJPU","outputId":"1a115cfd-bdd0-4863-87de-bd8eeda119a3"},"outputs":[{"output_type":"stream","name":"stdout","text":["MSA server enabled: https://api.colabfold.com\n","MSA server authentication: no credentials provided\n","Downloading the CCD data to /root/.boltz/mols.tar. This may take a bit of time. You may change the cache directory with the --cache flag.\n","Extracting the CCD data to /root/.boltz/mols. This may take a bit of time. You may change the cache directory with the --cache flag.\n","Downloading the Boltz-2 weights to /root/.boltz/boltz2_conf.ckpt. You may change the cache directory with the --cache flag.\n","Downloading the Boltz-2 affinity weights to /root/.boltz/boltz2_aff.ckpt. You may change the cache directory with the --cache flag.\n","Checking input data.\n","Processing 1 inputs with 1 threads.\n","  0% 0/1 [00:00<?, ?it/s]Generating MSA for /tmp/tmpz_cwhjmm.fasta with 1 protein entities.\n","Calling MSA server for target tmpz_cwhjmm with 1 sequences\n","MSA server URL: https://api.colabfold.com\n","MSA pairing strategy: greedy\n","No authentication provided for MSA server\n","\n","  0% 0/150 [00:00<?, ?it/s]\u001b[A\n","SUBMIT:   0% 0/150 [00:00<?, ?it/s]\u001b[A\n","COMPLETE:   0% 0/150 [00:00<?, ?it/s]\u001b[A\n","COMPLETE: 100% 150/150 [00:01<00:00, 77.76it/s] \n","100% 1/1 [00:02<00:00,  2.43s/it]\n","Traceback (most recent call last):\n","  File \"/usr/local/bin/boltz\", line 8, in <module>\n","    sys.exit(cli())\n","             ^^^^^\n","  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 1442, in __call__\n","    return self.main(*args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 1363, in main\n","    rv = self.invoke(ctx)\n","         ^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 1830, in invoke\n","    return _process_result(sub_ctx.command.invoke(sub_ctx))\n","                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 1226, in invoke\n","    return ctx.invoke(self.callback, **ctx.params)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 794, in invoke\n","    return callback(*args, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/content/boltz/src/boltz/main.py\", line 1256, in predict\n","    trainer = Trainer(\n","              ^^^^^^^^\n","  File \"/usr/local/lib/python3.11/dist-packages/pytorch_lightning/utilities/argparse.py\", line 70, in insert_env_defaults\n","    return fn(self, **kwargs)\n","           ^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.11/dist-packages/pytorch_lightning/trainer/trainer.py\", line 404, in __init__\n","    self._accelerator_connector = _AcceleratorConnector(\n","                                  ^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.11/dist-packages/pytorch_lightning/trainer/connectors/accelerator_connector.py\", line 144, in __init__\n","    self._accelerator_flag = self._choose_gpu_accelerator_backend()\n","                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n","  File \"/usr/local/lib/python3.11/dist-packages/pytorch_lightning/trainer/connectors/accelerator_connector.py\", line 354, in _choose_gpu_accelerator_backend\n","    raise MisconfigurationException(\"No supported gpu backend found!\")\n","lightning_fabric.utilities.exceptions.MisconfigurationException: No supported gpu backend found!\n"]}],"source":["# list of input FASTA strings, labelled as protein or dna as required by Boltz\n","fasta_list = [\">A|protein\\nSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNT\",\n","              \">B|dna\\nTTTCCTAGACTTGCCCAATTA\",\n","              \">C|dna\\nATAATTGGGCAAGTCTAGGAA\"]\n","\n","# make a tempfile to use as input for boltz to avoid having to generate unnecessary files\n","fasta_text = \"\\n\".join(fasta_list)\n","\n","with tempfile.NamedTemporaryFile(mode=\"w\", suffix=\".fasta\", delete=False) as fh:\n","    fh.write(fasta_text)\n","    fasta_path = pathlib.Path(fh.name)\n","\n","# get the filename from the output path to use as an output ID\n","output_id = fasta_path.name.split(\".\")[0]\n","\n","output_dir = pathlib.Path(\"boltz_output\")\n","\n","!boltz predict {fasta_path} --out_dir {output_dir} --use_msa_server\n"]},{"cell_type":"markdown","metadata":{"id":"Dwpo2tHt_O7o"},"source":["## Compare the prediction to the PDB entry\n","\n","In this case there is a PDB entry we can use as a reference and we can visualise this here and see that the prediction is quite accurate.\n","\n","First this is the PDB entry:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":617},"executionInfo":{"elapsed":35,"status":"ok","timestamp":1753103873360,"user":{"displayName":"Peter Canning","userId":"16600209944252973081"},"user_tz":-60},"id":"jxOJMW35-_4q","outputId":"4be973bf-4386-48fa-b418-037da51a540f"},"outputs":[{"output_type":"display_data","data":{"application/3dmoljs_load.v0":"<div id=\"3dmolviewer_1753103873309551\"  style=\"position: relative; width: 800px; height: 600px;\">\n        <p id=\"3dmolwarning_1753103873309551\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    //this is to ignore the existence of requirejs amd\n    var savedexports, savedmodule;\n    if (typeof exports !== 'undefined') savedexports = exports;\n    else exports = {}\n    if (typeof module !== 'undefined') savedmodule = module;\n    else module = {}\n\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n        exports = savedexports;\n        module = savedmodule;\n        resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://cdn.jsdelivr.net/npm/3dmol@2.5.1/build/3Dmol-min.js');\n}\n\nvar viewer_1753103873309551 = null;\nvar warn = document.getElementById(\"3dmolwarning_1753103873309551\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_1753103873309551 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_1753103873309551\"),{backgroundColor:\"white\"});\n$3Dmol.download(\"pdb:1TSR\", viewer_1753103873309551, {}, function() {\nviewer_1753103873309551.setStyle({\"cartoon\": {}});\nviewer_1753103873309551.zoomTo();\n\tviewer_1753103873309551.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\nviewer_1753103873309551.render();\n})\n});\n</script>","text/html":["<div id=\"3dmolviewer_1753103873309551\"  style=\"position: relative; width: 800px; height: 600px;\">\n","        <p id=\"3dmolwarning_1753103873309551\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n","        </div>\n","<script>\n","\n","var loadScriptAsync = function(uri){\n","  return new Promise((resolve, reject) => {\n","    //this is to ignore the existence of requirejs amd\n","    var savedexports, savedmodule;\n","    if (typeof exports !== 'undefined') savedexports = exports;\n","    else exports = {}\n","    if (typeof module !== 'undefined') savedmodule = module;\n","    else module = {}\n","\n","    var tag = document.createElement('script');\n","    tag.src = uri;\n","    tag.async = true;\n","    tag.onload = () => {\n","        exports = savedexports;\n","        module = savedmodule;\n","        resolve();\n","    };\n","  var firstScriptTag = document.getElementsByTagName('script')[0];\n","  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n","});\n","};\n","\n","if(typeof $3Dmolpromise === 'undefined') {\n","$3Dmolpromise = null;\n","  $3Dmolpromise = loadScriptAsync('https://cdn.jsdelivr.net/npm/3dmol@2.5.1/build/3Dmol-min.js');\n","}\n","\n","var viewer_1753103873309551 = null;\n","var warn = document.getElementById(\"3dmolwarning_1753103873309551\");\n","if(warn) {\n","    warn.parentNode.removeChild(warn);\n","}\n","$3Dmolpromise.then(function() {\n","viewer_1753103873309551 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_1753103873309551\"),{backgroundColor:\"white\"});\n","$3Dmol.download(\"pdb:1TSR\", viewer_1753103873309551, {}, function() {\n","viewer_1753103873309551.setStyle({\"cartoon\": {}});\n","viewer_1753103873309551.zoomTo();\n","\tviewer_1753103873309551.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\n","viewer_1753103873309551.render();\n","})\n","});\n","</script>"]},"metadata":{}}],"source":["# view the original PDB file:\n","view = py3Dmol.view(query='pdb:1TSR',style={'cartoon':{}}, width=800, height=600)\n","view.setStyle({\"cartoon\": {\"color\": \"spectrum\"}})\n","view.show()"]},{"cell_type":"markdown","metadata":{"id":"ufDKJJaHCmaq"},"source":["Now the prediction:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":365},"executionInfo":{"elapsed":21,"status":"error","timestamp":1753103873374,"user":{"displayName":"Peter Canning","userId":"16600209944252973081"},"user_tz":-60},"id":"vtCtS7ltCzek","outputId":"93ec5c5a-74f4-4da1-d18e-c41d4486ee4a"},"outputs":[{"output_type":"error","ename":"FileNotFoundError","evalue":"[Errno 2] No such file or directory: 'boltz_output/boltz_results_tmpz_cwhjmm/predictions/tmpz_cwhjmm/tmpz_cwhjmm_model_0.cif'","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m/tmp/ipython-input-5-2907248466.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mprediction_location\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0moutput_dir\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;34mf\"boltz_results_{output_id}\"\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;34m\"predictions\"\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0moutput_id\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;34mf\"{output_id}_model_0.cif\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mcif_str\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprediction_location\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_text\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mview\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpy3Dmol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mview\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mview\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maddModel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcif_str\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'mmcif'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.11/pathlib.py\u001b[0m in \u001b[0;36mread_text\u001b[0;34m(self, encoding, errors)\u001b[0m\n\u001b[1;32m   1056\u001b[0m         \"\"\"\n\u001b[1;32m   1057\u001b[0m         \u001b[0mencoding\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtext_encoding\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1058\u001b[0;31m         \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'r'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencoding\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1059\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1060\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.11/pathlib.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, mode, buffering, encoding, errors, newline)\u001b[0m\n\u001b[1;32m   1042\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;34m\"b\"\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1043\u001b[0m             \u001b[0mencoding\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtext_encoding\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1044\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffering\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencoding\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnewline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1045\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1046\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mread_bytes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'boltz_output/boltz_results_tmpz_cwhjmm/predictions/tmpz_cwhjmm/tmpz_cwhjmm_model_0.cif'"]}],"source":["# this is the location of the output predicted structure\n","prediction_location = output_dir / f\"boltz_results_{output_id}\" / \"predictions\" / output_id / f\"{output_id}_model_0.cif\"\n","\n","cif_str = prediction_location.read_text()\n","view = py3Dmol.view()\n","view.addModel(cif_str, 'mmcif')\n","view.setStyle({\"cartoon\": {\"color\": \"spectrum\"}})\n","view.zoomTo()\n","view.show()"]},{"cell_type":"markdown","metadata":{"id":"dtdTuntlqlxg"},"source":["## Design a binder\n","\n","We would like to design an artificial protein that will bind to the DNA-binding interface of p53. From the predicted structure we can identify the DNA binding interface and key residues involved in the protein:DNA interaction. The region around Ser241 seems to be a crucial interaction site with the DNA, so we will designate some 'hotspot' residues: Ser241, Ala276 and Asp281.\n","\n","On a real project you might want to probe the interface further with, for example, mutagenesis or molecular dynamics data.\n","\n","I designed a series of 10 binders (each 100-residues) that will bind to this interface using RFdiffusion using the command:\n","\n","`./scripts/run_inference.py inference.output_prefix=binder_outputs/p53_binder inference.input_pdb=input_pdbs/1tsr.pdb 'contigmap.contigs=[A94-289/0 100-100]' 'ppi.hotspot_res=[A241,A276,A281]' inference.num_designs=10`\n","\n","I haven't made this part of the notebook for simplicity - RFdiffusion is setup more as a CLI tool and takes a while to run without extra compute resources.\n","\n","I selected the last of these designs and used proteinMPNN to design sequences that would yield the desired structure (ProteinMPNN is also more of a CLI tool):\n","\n","`python protein_mpnn_run.py --pdb_path input_files/p53_binder_9.pdb --out_folder output_files --num_seq_per_target 10 --sampling_temp \"0.1\" --seed 0 --batch_size 1 --model_name v_48_020`\n","\n","This resulted in 10 binder sequences for this one design (it is possible to design many more binder structures and multiple sequences for each structure):"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"S2mgRwZ3r0k4"},"outputs":[],"source":["binder_sequences = [\"SLAALLAAAALAAAAAAAAARQPEADAAILAAVRAAVAALRAAIAAAAAAGVDREAAVAAARAMAAAVEATVARHAAACAPLAPEVAALAAEVAALAAEL\",\n","\"SAAAAAAAAAAAAAAAAAAARQPARDAEIQAEVAAAIAAVRAARAAAAAAGVDKAAAVAALKAAVDAVVATVAKHAAATAPLKAEVEAVRAEVDALAAAL\",\n","\"DAAAAAAAAAAAAAAAAAAAAAPARAAAIRAEVRAAIAALRELRKRAEEAGVDRAEAVAALAALAEAIAAAAAAHAAATAPLRPEIEAARAEVEALRAEL\",\n","\"SAAAAAAAAALAAAAAAAEAEAPARAAAIRAEVAAGIEELEEKEEEYEEEGVDEEEGVEELEEMMEEIVALAERHPEAVAPLKPRIEERRERVEELKEEL\",\n","\"SAAAAAAAAAAAAAAAAAEAAAPARCAAIQAEVRAAIAAIEAAIEQAEKEGVDPAAAIAAARAMVDEVRARAERHAACTAPLAPEIDAVAARVAARAAAA\",\n","\"DLAALLAAALAAALAAAAAAAAPARDAAILAELRERIAAIRELKAEAERAGVDRAAAVAALAAAMAEARALAAAHPAATAPLAPELDAVAAEVAALVEAL\",\n","\"SLAAALAAALAAAAAAAAAAAQPAADAAVRAAVAAAIAALRAARAAAAAAGVDPAAAVAALAAAADAVRAAAAKHPAAAAPLAPEIAAAAAAVAAEAAAL\",\n","\"SAAAAAAAAAAAAAAAAAAARQPEREAEIKEKLKKMIEEIKELIKKAKKEGIDKKKAVEEAEKKVEEAKELAEKHPEATEPLKEELEKVKKEVEELKKKL\",\n","\"SLLAALAAALAAAAAAAAAAALPARCAAIQEKVRKMIEELKEEIKRAKEEGVDKKEAVERAEKMMEEIVKLAEEHPECVAPLKPEIEKAKKEVEEMKEKL\",\n","\"SPLAALLAALAAALAAALAAAQPARDAAIQAALQAAIAATREAIAAAAAAGVDPAAAAAAAAAMVAAARALAARNPAACAPLQPELDAVAAEVDALVAAL\"]"]},{"cell_type":"markdown","metadata":{"id":"DCi1Ul5ltzoz"},"source":["## Analyse and select a design\n","\n","Lets make a multiple sequence alignment from these binder sequences:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"CbIaZ7fhtPuK"},"outputs":[],"source":["from Bio.Seq import Seq\n","from Bio.SeqRecord import SeqRecord\n","from Bio.Align import MultipleSeqAlignment\n","\n","alignment = MultipleSeqAlignment([SeqRecord(Seq(sequence), id=f\"binder{binder_sequences.index(sequence)}\") for sequence in binder_sequences])\n","print(alignment)"]},{"cell_type":"markdown","metadata":{"id":"sOYPTATGvT7T"},"source":["We can now identify the sequence with the highest solubility for this example. This can be a useful metric to select binders to test, perhaps coupled with other metrics such as stability.\n","\n","We are using a very simple measure of solubility, just counting the fraction of hydrophilic residues. For a more sophisticated analysis we could use a deep-learning framework such as [DeepSol](https://academic.oup.com/bioinformatics/article/34/15/2605/4938490)."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"23xF20bYvJfD"},"outputs":[],"source":["def simple_solubility_score(seq):\n","    hydrophilic_residues = \"DEHKR\"\n","    return sum(1 for aa in seq if aa in hydrophilic_residues) / len(seq)\n","\n","solubility_scores = {}\n","for sequence in binder_sequences:\n","  solubility_scores[sequence] = simple_solubility_score(sequence)\n","\n","print(solubility_scores)\n","\n","soluble_binder = max(solubility_scores, key=solubility_scores.get)\n","print(f\"The sequence with the highest solubility score is: {soluble_binder}\")"]},{"cell_type":"markdown","metadata":{"id":"vtz8fMamucHx"},"source":["In our example case, we are trying to do something quite difficult which is to design a binder to a charged polar site on p53 so I have taken all the sequences for this design and predicted the structure of the complex with p53 to identify one which binds at the desired interface.\n","\n","The selected binder sequence does not have great solubility but does appear to bind at the right site, including our specified hotspot residues. For a real project it might be preferable to leverage more compute power to design multiple binder structures and multiple sequences per design, then set up a workflow to systematically predict and analyse the bound structures and filter to those with preferred properties."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"K53AudqhuA-5"},"outputs":[],"source":["p53_and_binder = [fasta_list[0],\n","              f\">B|protein\\n{binder_sequences[5]}]\"]\n","\n","# make a tempfile to use as input for boltz to avoid having to generate unnecessary files\n","p53_and_binder_text = \"\\n\".join(p53_and_binder)\n","\n","with tempfile.NamedTemporaryFile(mode=\"w\", suffix=\".fasta\", delete=False) as fh:\n","    fh.write(p53_and_binder_text)\n","    fasta_path = pathlib.Path(fh.name)\n","\n","output_id = fasta_path.name.split(\".\")[0]\n","\n","output_dir = pathlib.Path(\"boltz_output\")\n","\n","!boltz predict {fasta_path} --out_dir {output_dir} --use_msa_server\n","\n","# this is the location of the output predicted structure\n","prediction_location = output_dir / f\"boltz_results_{output_id}\" / \"predictions\" / output_id / f\"{output_id}_model_0.cif\"\n","\n","print(\"loading prediction: \", prediction_location)\n","\n","cif_str = prediction_location.read_text()\n","view = py3Dmol.view()\n","view.addModel(cif_str, 'mmcif')\n","view.setStyle({\"cartoon\": {\"color\": \"spectrum\"}})\n","view.zoomTo()\n","view.show()"]}],"metadata":{"accelerator":"GPU","colab":{"gpuType":"T4","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}